fuse_core/include/fuse_core/serialization.h

```diff
@@ -0,0 +1,143 @@
+/*
+ * Software License Agreement (BSD License)
+ *
+ *  Copyright (c) 2019, Locus Robotics
+ *  All rights reserved.
+ *
+ *  Redistribution and use in source and binary forms, with or without
+ *  modification, are permitted provided that the following conditions
+ *  are met:
+ *
+ *   * Redistributions of source code must retain the above copyright
+ *     notice, this list of conditions and the following disclaimer.
+ *   * Redistributions in binary form must reproduce the above
+ *     copyright notice, this list of conditions and the following
+ *     disclaimer in the documentation and/or other materials provided
+ *     with the distribution.
+ *   * Neither the name of the copyright holder nor the names of its
+ *     contributors may be used to endorse or promote products derived
+ *     from this software without specific prior written permission.
+ *
+ *  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ *  "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ *  LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+ *  FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ *  COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ *  INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ *  BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ *  LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ *  CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ *  LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ *  ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ */
+#ifndef FUSE_CORE_SERIALIZATION_H
+#define FUSE_CORE_SERIALIZATION_H
+
+#include <fuse_core/uuid.h>
+
+#include <ros/time.h>
+
+#include <boost/archive/binary_iarchive.hpp>
+#include <boost/archive/binary_oarchive.hpp>
+#include <boost/archive/text_oarchive.hpp>
+#include <boost/archive/text_iarchive.hpp>
+#include <boost/serialization/access.hpp>
+#include <boost/serialization/array_wrapper.hpp>
+#include <boost/serialization/base_object.hpp>
+#include <boost/serialization/export.hpp>
+#include <boost/serialization/shared_ptr.hpp>
+#include <boost/uuid/uuid_serialize.hpp>
+#include <Eigen/Core>
+
+
+namespace fuse_core
+{
+  using BinaryInputArchive = boost::archive::binary_iarchive;
+  using BinaryOutputArchive = boost::archive::binary_oarchive;
+  using TextInputArchive = boost::archive::text_iarchive;
+  using TextOutputArchive = boost::archive::text_oarchive;
+}
+
+namespace boost
+{
+namespace serialization
+{
+
+/**
+ * @brief Serialize a ros::Time variable using Boost Serialization
+ */
+template<class Archive>
+void serialize(Archive& archive, ros::Time& stamp, const unsigned int /* version */)
+{
+  archive & stamp.sec;
+  archive & stamp.nsec;
+}
+
+/**
+ * @brief Serialize an Eigen Matrix using Boost Serialization
+ */
+template<class Archive,
```

Comment:
I think the correct way to take an Eigen matrix as a function argument is described here: https://eigen.tuxfamily.org/dox/TopicFunctionTakingEigenTypes.html

It'd also allow to pass Eigen expressions, not just a matrix, which I think it should be fine, and maybe actually useful in some cases.

There are two options:
1. Template function
2. `Eigen::Ref<>`

I think in this case we need the template version, which would look like this:
``` c++
template <class Archive, typename Derived>
void save(Archive& archive, const EIen::MatrixBase<Derived>& matrix, const unsigned int /* version */)
```

The same applies below for the `load` function.

However, if you're planning to have specializations for the 4 cases of `MaxRows_` and `MaxCols_` being fixed size or `Eigen::Dynamic`, then it's probably better to leave this as is now. What you have know works for all cases, but for a fixed size matrix you don't need to serialize the `rows` and `cols`. Calling `resize` on a fixed-size matrix, on the other size, would have no effect or throw an assert, which sounds good to me: https://eigen.tuxfamily.org/dox/group__TutorialMatrixClass.html#title6

Reply:
Yeah. This was one of those things I pulled from SO. I'm was a bit curious about the resizing thing for static-sized arrays, but it seems to work. It is something I plan to revisit at some future date, as much for curiosity as anything.

That said, I'm not sure that accepting a matrix by base type is the correct thing to do in this case. Since we are serializing, we need the full derived type. But the CRTP of MatrixBase may change things. I haven't thought it through.

There is an interesting suggestion here to use "Eigen's plugin feature"...which I didn't know existed.
https://stackoverflow.com/questions/18382457/eigen-and-boostserialize

Anyway, what is here I think is "good enough" for a first implementation. But if you want to make me a PR with better Eigen serialization support, that would be welcome.

Reply:
I completely agree this can be explored later.